### PR TITLE
bugfix: fix finished transaction swallows exception when committing

### LINKED ISSUE #5468: 全局事务回滚后，没有抛出任何异常
1679879376282_DE2EE361-A5A3-4426-B25A-54390984AF9C
以上时该全局事务的所有日志，Xid:192.168.0.144:8091:1297400804576108061

### PR DESCRIPTION
**Pull request by slievrly**
- [ ] I have registered the PR [changes](https://github.com/seata/seata/tree/develop/changes).

### Ⅰ. Describe what this PR did

fix finished transaction swallows exception when committing

### Ⅱ. Does this pull request fix one issue?

fix #5468

### CODE REVIEW

File: tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java [hunk 1]
```diff
@@ -214,19 +215,30 @@ private void commitTransaction(GlobalTransaction tx, TransactionInfo txInfo)
...
+                    code = TransactionalExecutor.Code.CommitFailure;
+                    break;
                 default:
             }
-            if (code != TransactionalExecutor.Code.Unknown) {
-                Exception exx = new TmTransactionException(TransactionExceptionCode.TransactionTimeout, String.format("Global transaction[%s] is timeout and will be rollback[TC].", tx.getXid()));
-                throw new TransactionalExecutor.ExecutionException(tx, exx, code);
+            Exception statusException = null;
+            if (GlobalStatus.isTwoPhaseHeuristic(afterCommitStatus)) {
+                statusException = new TmTransactionException(TransactionExceptionCode.CommitHeuristic,
```
Comment (a364176773) [inline tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java:234]: 这里如果是rollbacked应该是 may be TimeoutRollbacked吧?

Comment (slievrly) [reply tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java:234]: In the two-phase commit scenario, the transaction state is TimeoutRollbacked.

---

是指二阶段commit场景下的事务状态是TimeoutRollbacked场景嘛？

Comment (a364176773) [reply tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java:234]: 是的,就是这个日志是要表示最终是回滚还是说要表示事务最后状态,如果是最后的状态,当一个事务commit得到的却是finish,那肯定事务是被超时回滚了,因为同步决议时(非超时)只可能返回committ相关状态,如果是只是为了告诉用户事务可能已经回滚完了,那么这个日志看起来没问题.就是看最终表达上是偏向于结果,还是最有可能的状态

Comment (slievrly) [reply tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java:234]: The message that is communicated to the user is that the commit has failed. There are many reasons for this failure, such as storage exceptions, but these exceptions do not define a very fine state to pass to the user.
It's not necessarily the timeout rollback that causes a transaction to commit Finished, but it's definitely the biggest reason.

---

面向用户传递的信息就是提交失败，失败有很多原因比如存储异常，但是这种异常不会定义非常细的状态透传给用户。
事务提交Finished不一定是因为超时回滚导致，但超时回滚绝对是最大的原因。

Comment (a364176773) [reply tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java:234]: 明白了，那这个日志没问题

**Review by a364176773 — APPROVED** [review]
LGTM

File: tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java (hunk 1 above)
Comment (lbj812) [reply tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java:234]: 看你俩聊了半天，没看懂

**Pull request merged**